system/security/plugins/jwtSecurity/jwtSecurity.cpp

```diff
@@ -0,0 +1,1136 @@
+                    userInfo->addWUScopePerm(x->scopePattern, SecAccess_Full, defaultWUAccess, true);
+            }
+
+            // Add file scope permissions in alphabetical (and implicitly, length) order
+            std::sort(fileScopePerms.begin(), fileScopePerms.end());
+            for (CollectedScopePerms::const_iterator x = fileScopePerms.begin(); x != fileScopePerms.end(); x++)
+            {
+                if (x->claim == "AllowFileScopeAccess")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Access, defaultFileAccess);
+                else if (x->claim == "AllowFileScopeRead")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Read, defaultFileAccess);
+                else if (x->claim == "AllowFileScopeWrite")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Write, defaultFileAccess);
+                else if (x->claim == "AllowFileScopeFull")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Full, defaultFileAccess);
+                else if (x->claim == "DenyFileScopeAccess")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Access, defaultFileAccess, true);
+                else if (x->claim == "DenyFileScopeRead")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Read, defaultFileAccess, true);
+                else if (x->claim == "DenyFileScopeWrite")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Write, defaultFileAccess, true);
+                else if (x->claim == "DenyFileScopeFull")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Full, defaultFileAccess, true);
+            }
+
+            gUserCache.set(subject, userInfo);
+
+            verificationResult = true;
+        }
+        catch (const std::runtime_error& e)
+        {
+            // Most jwt-cpp exceptions are subclasses of std::runtime_error
+            PROGLOG("CJwtSecurityManager::verifyToken error: %s", e.what());
+        }
+        catch (const std::logic_error& e)
+        {
+            PROGLOG("CJwtSecurityManager::logic_error: %s", e.what());
+        }
+        catch (const std::bad_cast& e)
+        {
+            PROGLOG("CJwtSecurityManager::bad_cast error: %s", e.what());
+        }
+        catch (...)
+        {
+            PROGLOG("CJwtSecurityManager::verifyToken unknown error");
+            throw;
+        }
+
+        return verificationResult;
+    }
+
+    /**
+     * Load the contents of the key if necessary.  Contents loaded when
+     * at least KEY_FILE_CHECK_INTERVAL have passed since last check time
+     *
+     * Function will throw an exception on error.
+     *
+     * @return  true
+     */
+    bool ensureKeyLoaded()
+    {
+        if ((keyContentsTimer.elapsedMs() > KEY_FILE_CHECK_INTERVAL_MS) || keyContents.empty())
+        {
+            CriticalBlock block(crit);
+
```

Comment:
I don't think it would make a large difference here, but I've added the check.  It is a cheap check.  The contents should be changing only during extraordinary circumstances (key rotation, for example).

However:  This did highlight the need to protect access to the keyContents member during verifyToken() -- reading the contents when another thread is potentially updating the key would be a Bad Thing.